[9341] Implement HostnameEndpoint.__repr__

twm opened the pull request:
Implement `HostnameEndpoint.__repr__` which produces output like `<HostnameEndpoint example.com:80>`.

## Contributor Checklist:

* [x] There is an associated ticket in Trac. https://twistedmatrix.com/trac/ticket/9341
* [x] I have created a newsfragment in src/twisted/newsfragments/ (see: https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles )
* [x] I have updated the automated tests.

twm committed:
Implement HostnameEndpoint.__repr__

Signed-off-by: Gordon Bolt <gbolt@ciena.com>

twm committed:
Fix style issues in the tests

twm committed:
Wrap long lines

adiroiban reviewed (approved):
Looks good. Thanks a lot :)

Long time since I have not see a simple and nicely written PR :)

Maybe update the test to have explicit type checks.

This should make it explicit that the repr should be native str.

Python 2 is more relaxed when comparing unicode and bytes/str

```
Python 2.7.12 (default, Nov 20 2017, 18:23:56) 
>>> b'a' == u'a'
True

Python 3.5.2 (default, Nov 23 2017, 16:37:01) 
>>> b'a' == u'a'
False
```

adiroiban commented:
This is cleared to land, but before landing this make sure all test pass.

Buildbot is down (at least for my) so I guess that this is why tests are not reported http://buildbot.twistedmatrix.com/boxes-supported?branch=hostnameendpoint-repr-9341

twm commented:
Explicit type checks are a good idea. I will do that. Thank you for the review!

twm committed:
Add explicit type checks of repr strings

Be very, very specific that the result of __repr__ is a native string.

Now that there is more than one assertion, add vertical whitespace to
organize the test in arrange, act, assert form.

adiroiban commented:
Thanks. I have merged it as I was on it and I hoped I helped with writing the commit message :)

Let me know if you prefer to merge the PR yourself

twm commented:
@adiroiban No worries. Thank you!
